### ISSUE TITLE
WSGI error from `When` syntax

### ISSUE THREAD

**Issue by cecilia-donnelly** [OpenTechStrategies/streetcrm#338]
Hey, @pjsier, I merged #336 into master (thank you!) and deployed, only to get an error in the logs:

```
File "/var/www/swoptact/streetcrm/admin.py", line 357
     When(**event_count_dict, then='event__id'),
                            ^
] SyntaxError: invalid syntax
```

This works fine for me locally, but the server didn't like it.  Any ideas about this?  I rolled back to the earlier deployed version and everything seems fine.  That line was added in bc348108, as part of the reporting changes.  

I can add more of the logs if that would be helpful, but this seemed to be the relevant part.

**Comment by cecilia-donnelly** [comment]
(I had to roll back because the site was giving an Internal Server Error and not showing up, in case that wasn't clear.)

**Comment by pjsier** [comment]
@cecilia-donnelly just looked into this, and if I remember right the server is running Python 3.4 but we've been using 3.5 locally? My virtual environment is 3.5, and I just created a new one with 3.4 and it gets the same error. I think 3.4 handles kwargs differently, and it doesn't seem to like having `then` specified after unpacking `**event_count_dict`.

It works when I change it to:

```python
if event_count_dict:
  event_count_dict['then'] = 'event__id'
  event_count = Count(Case(
    When(**event_count_dict),
    output_field=IntegerField()
  ), distinct=True)
```

We'll probably want to upgrade to 3.5, but I can make that change. I've been using 3.5 locally though, so even though I know it runs with that fix on 3.4 I'd want to run through the other tests again to be sure. Do you think it makes more sense to upgrade or do that patch?

**Comment by cecilia-donnelly** [comment]
I think it makes more sense to upgrade python.  I didn't even realize we've been using 3.5 locally and 3.4 on the server, but of course you're right.  Good catch!  Thanks for the quick response.

**Comment by ots-ttm** [comment]
Patrick Sier <notifications@github.com> writes:
>@cecilia-donnelly just looked into this, and if I remember right the
>server is running Python 3.4 but we've been using 3.5 locally? My
>virtual environment is 3.5, and I just created a new one with 3.4 and
>it gets the same error. I think 3.4 handles kwargs differently, and
>it doesn't seem to like having then specified after unpacking
>**event_count_dict.
>
>It works when I change it to:
>
>if event_count_dict:
>  event_count_dict['then'] = 'event__id'
>  event_count = Count(Case(
>    When(**event_count_dict),
>    output_field=IntegerField()
>  ), distinct=True)
>
>We'll probably want to upgrade to 3.5, but I can make that change.
>I've been using 3.5 locally though, so even though I know it runs
>with that fix on 3.4 I'd want to run through the other tests again to
>be sure. Do you think it makes more sense to upgrade or do that
>patch?

I'm running 'master' (commit 3ec4f5e7) locally with Python 3.6 right now.  The application seems to work fine: I don't see any errors on my console, and things are working right so far in the browser.

But when I read https://docs.djangoproject.com/en/2.0/ref/models/conditional-expressions/ about the When expression, the examples there put the lookup dict last in the argument list.  Now, I would think that with all the rest of the args being keyword args, it shouldn't matter, but perhaps this is a Python 3.4 vs 3.5/3.6 difference.

I'd be curious to know what happens under Python 3.4 if we patch the original code by putting **event_count_dict second in the argument list.  I.e., this patch:

    index fc977ad..19a7b6d 100644
    --- streetcrm/admin.py
    +++ streetcrm/admin.py
    @@ -354,7 +354,7 @@ class STREETCRMAdminSite(admin.AdminSite):
                         event_count_dict[event_query] = query_dict[event_query]
                 if event_count_dict:
                     event_count = Count(Case(
    -                    When(**event_count_dict, then='event__id'),
    +                    When(then='event__id', **event_count_dict),
                         output_field=IntegerField()
                     ), distinct=True)
                 else:

Separately from the above, we should probably upgrade the server to 3.6 anyway, preferably via a dist-upgrade.  Cecilia, we can discuss that in our upcoming call.

Best regards,
-Karl

**Comment by kfogel** [comment]
I'm running 'master' (commit 3ec4f5e7) locally with Python 3.6 right now.  The application seems to work fine: I don't see any errors on my console, and things are working right so far in the browser.

But when I read https://docs.djangoproject.com/en/2.0/ref/models/conditional-expressions/ about the When expression, the examples there put the lookup dict last in the argument list.  Now, I would think that with all the rest of the args being keyword args, it shouldn't matter, but perhaps this is a Python 3.4 vs 3.5/3.6 difference.

I'd be curious to know what happens under Python 3.4 if we patch the original code by putting **event_count_dict second in the argument list.  I.e., this patch:

    index fc977ad..19a7b6d 100644
    --- streetcrm/admin.py
    +++ streetcrm/admin.py
    @@ -354,7 +354,7 @@ class STREETCRMAdminSite(admin.AdminSite):
                         event_count_dict[event_query] = query_dict[event_query]
                 if event_count_dict:
                     event_count = Count(Case(
    -                    When(**event_count_dict, then='event__id'),
    +                    When(then='event__id', **event_count_dict),
                         output_field=IntegerField()
                     ), distinct=True)
                 else:

Separately from the above, we should probably upgrade the server to 3.6 anyway, preferably via a dist-upgrade.  Cecilia, we can discuss that in our upcoming call.

**Comment by pjsier** [comment]
@kfogel That example you posted works in the 3.4 environment I set up to test the original bug, so it looks like it is related to ordering of `**kwargs` in 3.4. I can't seem to find the specific change in Python that caused it though. Upgrading to 3.6 after testing that out a bit makes sense to me though

**Comment by kfogel** [comment]
Thanks, @pjsier.  I would suggest that we switch the order anyway, now, and deploy that first.  That way our code will match the documentation for usage of `When()`.  _After_ establishing that that works, we can upgrade the server to Python 3.6 too, but we should regard that as a separate change.

(Or maybe this is what you were saying?)

**Comment by pjsier** [comment]
@kfogel I wasn't thinking of updating the order now, but that works, and I just opened #339 with that change

**Comment by kfogel** [comment]
Thanks, @pjsier!  About to have a call with Cecilia; we'll review/merge right after that (patch looks good to me, though).

**Comment by kfogel** [comment]
Closing because merged #339.
